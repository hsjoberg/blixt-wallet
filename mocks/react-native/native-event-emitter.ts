const EventEmitter = require('EventEmitter');
const RCTDeviceEventEmitter = require('RCTDeviceEventEmitter');

/**
 * Mock the NativeEventEmitter as a normal JS EventEmitter.
 */
export class NativeEventEmitter extends EventEmitter {
  constructor() {
    super(RCTDeviceEventEmitter.sharedSubscriber);
  }
}