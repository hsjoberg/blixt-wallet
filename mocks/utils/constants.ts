export const TLV_RECORD_NAME = 128100;

export const GITHUB_REPO_URL = "https://github.com/hsjoberg/blixt-wallet";
export const HAMPUS_EMAIL = "mailto:hampus.sjobergðŸ’©protonmail.com".replace("ðŸ’©", "@");
export const TELEGRAM = "https://t.me/blixtwallet";

export const PLATFORM = "android";

export const HEADER_MIN_HEIGHT = 53;
export const HEADER_MAX_HEIGHT = 195;
