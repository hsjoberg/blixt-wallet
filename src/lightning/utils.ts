import deepMapKeys from "deep-map-keys";

export const fixGrpcJsonResponse = <T>(response: { [keys: string]: any })  =>
  deepMapKeys<T>(response, (key) => {
    if (key[key.length - 1] === "_") {
      return key.slice(0, -1);
    }
    return "$" + key;
});
