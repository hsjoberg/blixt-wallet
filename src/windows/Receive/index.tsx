import React from "react"
import { createStackNavigator } from "@react-navigation/stack";

import ReceiveSetup from "./ReceiveSetup";
import ReceiveQr from "./ReceiveQr";
import { lnrpc } from "../../../proto/proto";

const Stack = createStackNavigator();

export type ReceiveStackParamList = {
  ReceiveSetup: undefined;
  ReceiveQr: {
    invoice: lnrpc.AddInvoiceResponse;
  };
}

export default () => {
  return (
    <Stack.Navigator initialRouteName="ReceiveSetup" screenOptions={{ headerShown: false, animationEnabled: false }}>
      <Stack.Screen name="ReceiveSetup" component={ReceiveSetup} />
      <Stack.Screen name="ReceiveQr" component={ReceiveQr} />
    </Stack.Navigator>
  )
}