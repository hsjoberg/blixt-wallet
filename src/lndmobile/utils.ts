import { NativeModules } from "react-native";
import deepMapKeys from "deep-map-keys";
import * as base64 from "base64-js";
import * as $protobuf from "protobufjs";
import { lnrpc } from "../../proto/proto";

const { LndMobile } = NativeModules;

export interface ISendRequestClass<IReq, Req> {
  create: (options: IReq) => Req;
  encode: (request: Req) => $protobuf.Writer;
}

export interface ISendResponseClass<Res> {
  decode: (reader: $protobuf.Reader | Uint8Array) => Res;
  toObject(message: Res, options?: $protobuf.IConversionOptions): { [k: string]: any };
}

export interface ISyncCommandOptions<IReq, Req, Res> {
    request: ISendRequestClass<IReq, Req>;
    response: ISendResponseClass<Res>;
    method: string;
    options: IReq;
}

export interface IStreamCommandOptions<IReq, Req> {
    request: ISendRequestClass<IReq, Req>;
    method: string;
    options: IReq;
}

export interface IStreamResultOptions<Res> {
  base64Result: string;
  response: ISendResponseClass<Res>;
}

/**
 * @throws
 */
export const sendCommand = async <IReq, Req, Res>({ request, response, method, options }: ISyncCommandOptions<IReq, Req, Res>): Promise<Res> => {
  try {
    const instance = request.create(options);
    const b64 = await LndMobile.sendCommand(method, base64.fromByteArray(request.encode(instance).finish()));
    return response.decode(base64.toByteArray(b64.data));
  } catch (e) { throw e; }
};

export const sendStreamCommand = async <IReq, Req>({ request, method, options }: IStreamCommandOptions<IReq, Req>, streamOnlyOnce: boolean = false): Promise<string> =>  {
  const instance = request.create(options);
  const response = await LndMobile.sendStreamCommand(method, base64.fromByteArray(request.encode(instance).finish()), streamOnlyOnce);
  return response;
};

export const decodeStreamResult = <Res>( { base64Result, response }: IStreamResultOptions<Res>): Res => {
  return response.decode(base64.toByteArray(base64Result));
};
