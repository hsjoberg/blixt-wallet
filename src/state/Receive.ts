import { EmitterSubscription } from "react-native";
import { Action, action, Thunk, thunk } from "easy-peasy";
import Long from "long";

import { IStoreModel } from "./index";
import { IStoreInjections } from "./store";
import { ITransaction } from "../storage/database/transaction";
import { lnrpc, invoicesrpc } from "../../proto/lightning";
import { setupDescription } from "../utils/NameDesc";
import { valueFiat, formatBitcoin } from "../utils/bitcoin-units";
import {
  timeout,
  decodeTLVRecord,
  bytesToHexString,
  toast,
  uint8ArrayToUnicodeString,
} from "../utils";
import { TLV_RECORD_NAME, TLV_SATOGRAM, TLV_WHATSAT_MESSAGE } from "../utils/constants";
import { identifyService } from "../utils/lightning-services";
import { LndMobileEventEmitter } from "../utils/event-listener";
import { checkLndStreamErrorResponse } from "../utils/lndmobile";
import { ILNUrlPayResponsePayerData } from "./LNURL";

import logger from "./../utils/log";
const log = logger("Receive");

// TODO(hsjoberg): this should match Transaction model
interface IInvoiceTempData {
  rHash?: string;
  payer: string | null;
  type: ITransaction["type"];
  website: string | null;
  callback?: (pr: string) => void;
  lightningBox?: {
    descHash: Uint8Array;
    payerData?: ILNUrlPayResponsePayerData;
  };
}

interface IReceiveModelAddInvoicePayload {
  description: string;
  sat: number;
  expiry?: number;
  tmpData?: IInvoiceTempData;
  preimage?: Uint8Array;
  skipNameDesc?: boolean;
}

interface IReceiveModelAddInvoiceBlixtLspPayload {
  description: string;
  sat: number;
  expiry?: number;
  tmpData?: IInvoiceTempData;

  preimage: Uint8Array;
  chanId: string;
  cltvExpiryDelta: number;
  feeBaseMsat: number;
  feeProportionalMillionths: number;
  servicePubkey: string;
}

interface IReceiveModelCancelInvoicePayload {
  rHash: string;
}

export interface IReceiveModel {
  initialize: Thunk<IReceiveModel, void, IStoreInjections>;
  deinitialize: Thunk<IReceiveModel>;

  addInvoice: Thunk<
    IReceiveModel,
    IReceiveModelAddInvoicePayload,
    IStoreInjections,
    IStoreModel,
    Promise<lnrpc.AddInvoiceResponse>
  >;
  addInvoiceBlixtLsp: Thunk<
    IReceiveModel,
    IReceiveModelAddInvoiceBlixtLspPayload,
    IStoreInjections,
    IStoreModel,
    Promise<lnrpc.AddInvoiceResponse>
  >;
  cancelInvoice: Thunk<
    IReceiveModel,
    IReceiveModelCancelInvoicePayload,
    IStoreInjections,
    Promise<invoicesrpc.CancelInvoiceResp>
  >;
  subscribeInvoice: Thunk<IReceiveModel, void, IStoreInjections, IStoreModel>;
  setInvoiceSubscriptionStarted: Action<IReceiveModel, boolean>;
  setInvoiceSubscriptionResource: Action<IReceiveModel, EmitterSubscription | undefined>;

  setInvoiceTmpData: Action<IReceiveModel, IInvoiceTempData>;
  deleteInvoiceTmpData: Action<IReceiveModel, string>;

  invoiceSubscriptionStarted: boolean; // TODO make computed value
  invoiceSubscription?: EmitterSubscription;
  invoiceTempData: { [key: string]: IInvoiceTempData };
}

export const receive: IReceiveModel = {
  initialize: thunk(async (actions, _, { getState, injections }) => {
    if (getState().invoiceSubscriptionStarted) {
      log.w("Receive.initialize() called when subscription already started");
      return;
    }
    const subscribeInvoices = injections.lndMobile.wallet.subscribeInvoices;
    await subscribeInvoices();
    await timeout(2000); // Wait for the stream to get ready
    actions.subscribeInvoice();
    return true;
  }),

  deinitialize: thunk((actions, _, { getState }) => {
    const invoiceSubscription = getState().invoiceSubscription;
    if (invoiceSubscription) {
      log.i("Removing invoice subscription");
      invoiceSubscription.remove();
      actions.setInvoiceSubscriptionStarted(false);
      actions.setInvoiceSubscriptionResource(undefined);
    }
  }),

  addInvoice: thunk(async (actions, payload, { injections, getStoreState, getStoreActions }) => {
    log.d("addInvoice()");
    const addInvoice = injections.lndMobile.index.addInvoice;
    const name = getStoreState().settings.name;
    const invoiceExpiry = getStoreState().settings.invoiceExpiry;
    const description = payload.skipNameDesc
      ? payload.description
      : setupDescription(payload.description, name);

    const result = await addInvoice(
      payload.sat,
      description,
      payload.expiry ?? invoiceExpiry,
      payload.tmpData?.lightningBox?.descHash,
      payload.preimage,
    );
    log.d("addInvoice() result", [result]);
    getStoreActions().clipboardManager.addToInvoiceCache(result.paymentRequest);

    if (payload.tmpData) {
      actions.setInvoiceTmpData({
        ...payload.tmpData,
        rHash: bytesToHexString(result.rHash),
      });
    }

    return result;
  }),

  addInvoiceBlixtLsp: thunk(
    async (actions, payload, { injections, getStoreState, getStoreActions }) => {
      log.d("addInvoice()");
      const addInvoiceBlixtLsp = injections.lndMobile.index.addInvoiceBlixtLsp;
      const name = getStoreState().settings.name;
      const description = setupDescription(payload.description, name);

      const result = await addInvoiceBlixtLsp({
        amount: payload.sat,
        preimage: payload.preimage,
        chanId: payload.chanId,
        cltvExpiryDelta: payload.cltvExpiryDelta,
        feeBaseMsat: payload.feeBaseMsat,
        feeProportionalMillionths: payload.feeProportionalMillionths,
        memo: description,
        servicePubkey: payload.servicePubkey,
      });

      // (payload.sat, description, payload.expiry);
      log.d("addInvoice() result", [result]);
      getStoreActions().clipboardManager.addToInvoiceCache(result.paymentRequest);

      payload.tmpData = payload.tmpData ?? {
        type: "DUNDER_ONDEMANDCHANNEL",
        payer: null,
        website: null,
      };

      if (payload.tmpData) {
        actions.setInvoiceTmpData({
          ...payload.tmpData,
          rHash: bytesToHexString(result.rHash),
          type: "DUNDER_ONDEMANDCHANNEL",
        });
      }

      return result;
    },
  ),

  cancelInvoice: thunk(async (_, payload, { injections }) => {
    const cancelInvoice = injections.lndMobile.index.cancelInvoice;
    return await cancelInvoice(payload.rHash);
  }),

  subscribeInvoice: thunk(
    (actions, _2, { getState, dispatch, injections, getStoreState, getStoreActions }) => {
      const decodePayReq = injections.lndMobile.index.decodePayReq;
      const decodeInvoiceResult = injections.lndMobile.wallet.decodeInvoiceResult;
      if (getState().invoiceSubscriptionStarted) {
        log.d("Receive.subscribeInvoice() called when subscription already started");
        return;
      }
      const invoiceSubscription = LndMobileEventEmitter.addListener(
        "SubscribeInvoices",
        async (e: any) => {
          try {
            log.i("New invoice event");
            const error = checkLndStreamErrorResponse("SubscribeInvoices", e);
            if (error === "EOF") {
              return;
            } else if (error) {
              log.d("Got error from SubscribeInvoices", [error]);
              throw error;
            }

            const invoice = decodeInvoiceResult(e.data);

            const rHash = bytesToHexString(invoice.rHash);
            log.d("invoice", [rHash]);

            const paymentRequest = invoice.paymentRequest
              ? await decodePayReq(invoice.paymentRequest)
              : undefined;

            if (invoice.isKeysend) {
              if (invoice.state !== lnrpc.Invoice.InvoiceState.SETTLED) {
                log.i("Found keysend payment, but waiting for invoice state SETTLED");
                return;
              }
            }

            // TODO in the future we should handle
            // both value (the requested amount in the payreq)
            // and amtPaidMsat (the actual amount paid)
            if (!Long.isLong(invoice.value)) {
              invoice.value = Long.fromValue(invoice.value);
            }
            if (!Long.isLong(invoice.valueMsat)) {
              invoice.amtPaidSat = Long.fromValue(invoice.amtPaidSat);
            }
            if (!Long.isLong(invoice.amtPaidMsat)) {
              invoice.amtPaidMsat = Long.fromValue(invoice.amtPaidMsat);
            }

            const tmpData: IInvoiceTempData = getState().invoiceTempData[rHash] ?? {
              rHash,
              payer: null,
              type: "NORMAL",
              website: null,
            };

            const transaction: ITransaction = {
              description: invoice.memo || (invoice.isKeysend ? "Keysend payment" : ""),
              value: invoice.value ?? Long.fromInt(0),
              valueMsat: invoice.value.mul(1000) ?? Long.fromInt(0),
              amtPaidSat: invoice.amtPaidSat,
              amtPaidMsat: invoice.amtPaidMsat,
              fee: null,
              feeMsat: null,
              date: invoice.creationDate,
              duration: 0,
              expire: paymentRequest
                ? invoice.creationDate.add(paymentRequest.expiry)
                : Long.fromNumber(0),
              remotePubkey: paymentRequest ? paymentRequest.destination : "", // TODO
              status: decodeInvoiceState(invoice.state),
              paymentRequest: invoice.paymentRequest,
              rHash,
              nodeAliasCached: null,
              valueUSD: null,
              valueFiat: null,
              valueFiatCurrency: null,
              tlvRecordName: null,
              lightningAddress: null,
              lud16IdentifierMimeType: null,

              payer: tmpData.payer,
              website: tmpData.website,
              identifiedService: identifyService(null, "", tmpData.website),

              type: tmpData.type,
              locationLat: null,
              locationLong: null,

              preimage: invoice.rPreimage,
              lnurlPayResponse: null,
              lud18PayerData: tmpData.lightningBox?.payerData ?? null,

              hops: [],
            };

            if (invoice.state === lnrpc.Invoice.InvoiceState.SETTLED) {
              const fiatUnit = getStoreState().settings.fiatUnit;
              const valFiat = valueFiat(invoice.amtPaidSat, getStoreState().fiat.currentRate);

              transaction.valueUSD = valueFiat(
                invoice.amtPaidSat,
                getStoreState().fiat.fiatRates.USD.last,
              );
              transaction.valueFiat = valFiat;
              transaction.valueFiatCurrency = fiatUnit;

              let tlvRecordWhatSatMessage: string | undefined = undefined;
              let isSatogram = false;
              // Loop through known TLV records
              for (const htlc of invoice.htlcs) {
                if (htlc.customRecords) {
                  for (const [customRecordKey, customRecordValue] of Object.entries(
                    htlc.customRecords,
                  )) {
                    const decodedTLVRecord = decodeTLVRecord(customRecordKey);
                    if (decodedTLVRecord === TLV_RECORD_NAME) {
                      const tlvRecordName = uint8ArrayToUnicodeString(customRecordValue);
                      log.i("Found TLV_RECORD_NAME 🎉", [tlvRecordName]);
                      transaction.tlvRecordName = tlvRecordName;
                    } else if (decodedTLVRecord === TLV_WHATSAT_MESSAGE) {
                      tlvRecordWhatSatMessage = uint8ArrayToUnicodeString(customRecordValue);
                      log.i("Found TLV_WHATSAT_MESSAGE 🎉", [tlvRecordWhatSatMessage]);
                      if (invoice.isKeysend) {
                        transaction.description = tlvRecordWhatSatMessage;
                      }
                    } else if (decodedTLVRecord === TLV_SATOGRAM) {
                      log.i("Got a Satogram");
                      isSatogram = true;
                    } else {
                      log.i("Unknown TLV record", [decodedTLVRecord]);
                    }
                  }
                }
              }

              const bitcoinUnit = getStoreState().settings.bitcoinUnit;
              let message = `Received ${formatBitcoin(
                invoice.amtPaidSat,
                bitcoinUnit,
              )} (${valFiat.toFixed(2)} ${fiatUnit})`;
              if (transaction.tlvRecordName ?? transaction.payer ?? transaction.website) {
                message += ` from ${
                  transaction.tlvRecordName ?? transaction.payer ?? transaction.website
                }`;
              }
              if (tlvRecordWhatSatMessage) {
                message += ` with the message: ` + tlvRecordWhatSatMessage;
              }

              if (!isSatogram) {
                getStoreActions().notificationManager.localNotification({
                  message,
                  importance: "high",
                });
              }

              // We can now delete the temp data
              // as the invoice has been settled
              actions.deleteInvoiceTmpData(rHash);
            } else if (invoice.state === lnrpc.Invoice.InvoiceState.OPEN) {
              if (tmpData.callback) {
                tmpData.callback(invoice.paymentRequest);
              }
            }

            if (
              transaction.payer === "Hampus Sjöberg" ||
              transaction.payer === "Hampus Sjoberg" ||
              transaction.tlvRecordName === "Hampus Sjöberg" ||
              transaction.tlvRecordName === "Hampus Sjoberg"
            ) {
              transaction.identifiedService = "hampus";
            }

            setTimeout(async () => {
              await dispatch.channel.getBalance();
              await dispatch.fiat.getRate();
            }, 500);
            await dispatch.transaction.syncTransaction(transaction);
          } catch (error) {
            log.e("Error receiving invoice", [error]);
            toast("Error receiving payment: " + error.message, undefined, "danger");
          }
        },
      );
      actions.setInvoiceSubscriptionStarted(true);
      actions.setInvoiceSubscriptionResource(invoiceSubscription);
      log.i("Transaction subscription started");
    },
  ),

  setInvoiceSubscriptionStarted: action((state, payload) => {
    state.invoiceSubscriptionStarted = payload;
  }),

  setInvoiceSubscriptionResource: action((state, payload) => {
    state.invoiceSubscription = payload;
  }),

  setInvoiceTmpData: action((state, payload) => {
    state.invoiceTempData = {
      ...state.invoiceTempData,
      [payload.rHash!]: payload,
    };
  }),

  deleteInvoiceTmpData: action((state, payload) => {
    const tmpData = { ...state.invoiceTempData };
    delete tmpData[payload];
    state.invoiceTempData = tmpData;
  }),

  invoiceSubscriptionStarted: false,
  invoiceTempData: {},
};

function decodeInvoiceState(invoiceState: lnrpc.Invoice.InvoiceState) {
  switch (invoiceState) {
    case lnrpc.Invoice.InvoiceState.ACCEPTED:
      return "ACCEPTED";
    case lnrpc.Invoice.InvoiceState.CANCELED:
      return "CANCELED";
    case lnrpc.Invoice.InvoiceState.OPEN:
      return "OPEN";
    case lnrpc.Invoice.InvoiceState.SETTLED:
      return "SETTLED";
    default:
      return "UNKNOWN";
  }
}
