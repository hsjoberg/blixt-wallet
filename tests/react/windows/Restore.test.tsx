import React from "react";
import { render } from "@testing-library/react-native";
import { StoreProvider } from "easy-peasy";
import Long from "long";

import Receive from "../../../src/windows/Welcome/Restore";
import { createNavigationContainer, setupStore, setDefaultAsyncStorage, initCommonStore } from "../../utils";

jest.setTimeout(10000);

const AppContainer = createNavigationContainer(Receive, "Receive");

it("renders correctly", () => {
  const store = setupStore();

  const { unmount, toJSON } = render(
    <StoreProvider store={store}>
      {AppContainer}
    </StoreProvider>
  );
  expect(toJSON()).toMatchSnapshot();

  unmount();
});