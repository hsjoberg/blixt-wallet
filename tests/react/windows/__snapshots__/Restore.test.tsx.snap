// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={true}
      removeClippedSubviews={false}
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "overflow": "hidden",
          },
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": Array [
              Object {
                "translateY": 0,
              },
            ],
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
                "marginTop": 64,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "#fff",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "backgroundColor": "#151314",
                            "flex": 1,
                            "height": "100%",
                            "width": "100%",
                          },
                          Object {},
                          undefined,
                        ]
                      }
                    >
                      <ScrollView
                        automaticallyAdjustContentInsets={false}
                        contentContainerStyle={
                          Array [
                            Object {
                              "padding": undefined,
                            },
                            Array [
                              Object {
                                "padding": 14,
                              },
                              Object {
                                "alignItems": "center",
                                "flex": 1,
                                "justifyContent": "center",
                                "padding": 24,
                                "width": "100%",
                              },
                            ],
                          ]
                        }
                        contentInset={
                          Object {
                            "bottom": 0,
                          }
                        }
                        enableAutomaticScroll={true}
                        enableOnAndroid={false}
                        enableResetScrollToCoords={true}
                        extraHeight={75}
                        extraScrollHeight={0}
                        keyboardDismissMode="interactive"
                        keyboardOpeningTime={250}
                        keyboardShouldPersistTaps="handled"
                        keyboardSpace={0}
                        resetScrollToCoords={
                          Object {
                            "x": 0,
                            "y": 0,
                          }
                        }
                        scrollEventThrottle={1}
                        showsVerticalScrollIndicator={true}
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "flex": 1,
                          }
                        }
                        viewIsInsideTabBar={false}
                      >
                        <View
                          style={
                            Array [
                              Object {},
                              Object {
                                "height": "100%",
                                "width": "100%",
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {},
                                Object {
                                  "alignContent": "center",
                                  "flexDirection": "column",
                                  "height": 300,
                                  "justifyContent": "center",
                                  "marginTop": 32,
                                },
                              ]
                            }
                          >
                            <TextInput
                              allowFontScaling={true}
                              bordered={false}
                              editable={true}
                              multiline={true}
                              placeholderTextColor="#575757"
                              rejectResponderTermination={true}
                              rowSpan={6}
                              style={
                                Object {
                                  "backgroundColor": "#232323",
                                  "color": "#000",
                                  "fontSize": 20,
                                  "height": 150,
                                  "paddingLeft": 10,
                                  "paddingRight": 5,
                                  "textAlignVertical": "top",
                                  "width": "100%",
                                }
                              }
                              underline={false}
                              underlineColorAndroid="rgba(0,0,0,0)"
                              value=""
                            />
                            <View
                              style={
                                Array [
                                  Object {},
                                  Object {
                                    "display": "flex",
                                    "marginTop": 14,
                                    "width": "100%",
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  Object {
                                    "color": "#000",
                                    "fontSize": 21,
                                    "lineHeight": 22,
                                  }
                                }
                              >
                                Channel backup
                              </Text>
                              <View
                                style={
                                  Array [
                                    Object {},
                                    Object {
                                      "display": "flex",
                                      "flexDirection": "column",
                                    },
                                  ]
                                }
                              >
                                <TouchableOpacity
                                  activeOpacity={0.5}
                                  small={true}
                                  style={
                                    Object {
                                      "alignItems": "center",
                                      "backgroundColor": "#007aff",
                                      "borderBottomWidth": null,
                                      "borderColor": "#007aff",
                                      "borderLeftWidth": null,
                                      "borderRadius": 5,
                                      "borderRightWidth": null,
                                      "borderTopWidth": null,
                                      "borderWidth": undefined,
                                      "elevation": 2,
                                      "flexDirection": "row",
                                      "height": 30,
                                      "justifyContent": "space-between",
                                      "marginBottom": 10,
                                      "marginTop": 6,
                                      "paddingBottom": 6,
                                      "paddingTop": 6,
                                      "shadowColor": undefined,
                                      "shadowOffset": undefined,
                                      "shadowOpacity": undefined,
                                      "shadowRadius": undefined,
                                    }
                                  }
                                >
                                  <Text
                                    style={
                                      Object {
                                        "backgroundColor": "transparent",
                                        "color": "#fff",
                                        "fontFamily": "System",
                                        "fontSize": 14,
                                        "marginLeft": 0,
                                        "marginRight": 0,
                                        "paddingLeft": 16,
                                        "paddingRight": 16,
                                      }
                                    }
                                    uppercase={false}
                                  >
                                    Choose channel backup file on disk
                                  </Text>
                                </TouchableOpacity>
                                <TouchableOpacity
                                  activeOpacity={0.5}
                                  small={true}
                                  style={
                                    Object {
                                      "alignItems": "center",
                                      "backgroundColor": "#007aff",
                                      "borderBottomWidth": null,
                                      "borderColor": "#007aff",
                                      "borderLeftWidth": null,
                                      "borderRadius": 5,
                                      "borderRightWidth": null,
                                      "borderTopWidth": null,
                                      "borderWidth": undefined,
                                      "elevation": 2,
                                      "flexDirection": "row",
                                      "height": 30,
                                      "justifyContent": "space-between",
                                      "paddingBottom": 6,
                                      "paddingTop": 6,
                                      "shadowColor": undefined,
                                      "shadowOffset": undefined,
                                      "shadowOpacity": undefined,
                                      "shadowRadius": undefined,
                                    }
                                  }
                                >
                                  <Text
                                    style={
                                      Object {
                                        "backgroundColor": "transparent",
                                        "color": "#fff",
                                        "fontFamily": "System",
                                        "fontSize": 14,
                                        "marginLeft": 0,
                                        "marginRight": 0,
                                        "paddingLeft": 16,
                                        "paddingRight": 16,
                                      }
                                    }
                                    uppercase={false}
                                  >
                                    Restore via Google Drive
                                  </Text>
                                </TouchableOpacity>
                              </View>
                            </View>
                          </View>
                          <View
                            style={
                              Array [
                                Object {},
                                Object {
                                  "flex": 1,
                                  "marginTop": 24,
                                  "width": "100%",
                                },
                              ]
                            }
                          >
                            <Text
                              style={
                                Array [
                                  Object {
                                    "color": "#000",
                                    "fontSize": 27,
                                    "lineHeight": 32,
                                  },
                                  Object {
                                    "marginBottom": 3,
                                  },
                                ]
                              }
                            >
                              Restore wallet
                            </Text>
                            <Text
                              style={
                                Object {
                                  "color": "#000",
                                  "fontFamily": "System",
                                  "fontSize": 16,
                                }
                              }
                              uppercase={false}
                            >
                              To restore your wallet, write each word from your seed separated by a space.
                              

                              

                              If you wish to restore your Lightning Network channels, you need to provide a file backup.
                            </Text>
                          </View>
                        </View>
                        <View
                          style={
                            Array [
                              Object {},
                              Object {
                                "bottom": 24,
                                "width": "100%",
                              },
                            ]
                          }
                        >
                          <TouchableOpacity
                            activeOpacity={0.5}
                            block={true}
                            disabled={false}
                            style={
                              Object {
                                "alignItems": "center",
                                "alignSelf": "stretch",
                                "backgroundColor": "#007aff",
                                "borderBottomWidth": null,
                                "borderColor": "#007aff",
                                "borderLeftWidth": null,
                                "borderRadius": 5,
                                "borderRightWidth": null,
                                "borderTopWidth": null,
                                "borderWidth": undefined,
                                "elevation": 2,
                                "flexDirection": "row",
                                "height": 45,
                                "justifyContent": "center",
                                "paddingBottom": 6,
                                "paddingTop": 6,
                                "shadowColor": undefined,
                                "shadowOffset": undefined,
                                "shadowOpacity": undefined,
                                "shadowRadius": undefined,
                              }
                            }
                          >
                            <Text
                              style={
                                Object {
                                  "backgroundColor": "transparent",
                                  "color": "#fff",
                                  "fontFamily": "System",
                                  "fontSize": 16.5,
                                  "marginLeft": 0,
                                  "marginRight": 0,
                                  "paddingLeft": 16,
                                  "paddingRight": 16,
                                }
                              }
                              uppercase={false}
                            >
                              Restore Wallet
                            </Text>
                          </TouchableOpacity>
                        </View>
                      </ScrollView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      pointerEvents="box-none"
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        pointerEvents="box-none"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "borderBottomColor": "#a7a7aa",
                "flex": 1,
                "shadowColor": "#a7a7aa",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 18,
                  "opacity": 1,
                }
              }
            >
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "color": "rgba(0, 0, 0, 0.9)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                Receive
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
